### ISSUE TITLE
SA1600 should not trigger for explicit interface implementations

### ISSUE THREAD

**Issue by AArnott** [DotNetAnalyzers/StyleCopAnalyzers#3717]
Given the following declarations:

```cs
public interface IFoo
{
	/// <summary>
	/// My decoding method.
	/// </summary>
	static abstract bool TryDecode();
}

public class Foo : IFoo
{
	static bool IFoo.TryDecode() => true;
}
```

SA1600 flags the method on the `Foo` class.

This is unreasonable, since there is no way at all to invoke this method except by the interface, which is already documented. Adding `/// <inheritdoc />` to the implementation adds nothing. 

Per https://github.com/dotnet/roslyn/issues/70456, Roslyn does not offer a code fix to add the inheritdoc in the explicit interface implementation case, and this reasoning came up in the discussion.

**Comment by bjornhellander** [comment]
I am having a look at this. There seems to be quite a lot of dependencies between the SA16xx rules at the code level, with many of them calling a method inside the SA1600 analyzer. I need to try this out some more, but my gut says that this is not the only rule that should be affected by this issue. What is actually the desired behaviour? Should we treat explicitly implemented methods etc as if they were private?

**Comment by AArnott** [comment]
Thanks for looking, @bjornhellander.
I think the doc rules should _allow_ for the user not documenting explicit interface implementations. But if they _are_ present, the rules that ensure that docs follow certain conventions are probably still good to run. For example, if I _did_ document the method (for whatever reason) but left out the docs for one particular parameter, or put them out of order, then stylecop calling that out is arguably still a good thing.

**Comment by Alxandr** [comment]
SA1648 has the same issue. I get an error telling me that I can't use inheritdoc when doing `static TAggregate IAggregate<TAggregate, TEvent>.New(Guid id)`. Not sure if this qualifies as a new issue or not.

**Comment by bjornhellander** [comment]
@Alxandr Not the same problem. Yours is a duplicate of #3595 which has been fixed but not yet released.

**Comment by bjornhellander** [comment]
The following rules share code today that determine if the rule is applicable for a specific element, based on accessibility and documentation settings:
* SA1600 ElementsMustBeDocumented
* SA1601 PartialElementsMustBeDocumented
* _SA1602 EnumerationItemsMustBeDocumented (not relevant for this discussion)_
* SA1604 ElementDocumentationMustHaveSummary
* SA1605 PartialElementDocumentationMustHaveSummary
* SA1609 PropertyDocumentationMustHaveValue
* SA1611 PropertyDocumentationMustHaveValueText
* SA1612 ElementParameterDocumentationMustMatchElementParameters
* SA1615 ElementReturnValueMustBeDocumented
* SA1618 GenericTypeParametersMustBeDocumented
* SA1619 GenericTypeParametersMustBeDocumentedPartialClass

So for example if the element is private and it is not configured that private elements should be documented, then SA1600 does not trigger, but it also means that none of the other trigger, which they would have done by default if the element had been public.

Is the intended behaviour that none of the above should ever trigger for explicitly implemented members, regardless of documentation settings?

**Comment by AArnott** [comment]
> Is the intended behaviour that none of the above should ever trigger for explicitly implemented members, regardless of documentation settings?

IMO, no. I think SA1600 should not apply to explicit interface implementations. But the rest (assuming they only run when documentation is present) should run. The high level is: docs aren't required, but if they exist, they should follow the regular rules.

**Comment by bjornhellander** [comment]
> IMO, no. I think SA1600 should not apply to explicit interface implementations. But the rest (assuming they only run when documentation is present) should run. The high level is: docs aren't required, but if they exist, they should follow the regular rules.

Thanks for the input! The connections between the rules are intricate and my thinking might be flawed, but this is my take on it:

That would mean that SA1615 would require a return tag for a method if there is any documentation at all. For private members, that is not required by default. I don't think we should put higher demands on explicitly implemented members than on private ones, rather the contrary since explicit ones can never be used directly.

I definitely agree with you on rules that trigger on bad or missing text inside a tag (for example SA1616 for the returns tag), but those are other rules than the ones I listed above. (SA1611 looks suspicious though, I will have to check that one again.)
